p4-farm-nft/src/lib.rs

```diff
@@ -0,0 +1,2423 @@
+            farm_token_rates: to_U128s(&self.farm_token_rates),
+            farm_deposits: to_U128s(&self.farm_deposits),
+            is_active: self.is_active,
+            farming_start: self.farming_start,
+            farming_end: self.farming_end,
+            boost_nft_contracts: self.boost_nft_contracts.clone(),
+            total_staked: to_U128s(&self.total_stake),
+            total_farmed: to_U128s(&self.total_harvested),
+            total_boost: to_U128s(&self.total_boost),
+            fee_rate: self.fee_rate.into(),
+            accounts_registered: self.accounts_registered,
+            cheddar_rate: U128(self.cheddar_rate),
+            cheddar: self.cheddar.clone()
+        }
+    }
+
+    pub fn status(&self, account_id: AccountId) -> Option<P4Status> {
+        return match self.vaults.get(&account_id) {
+            Some(mut v) => {
+                let r = self.current_round();
+                v.ping(self.compute_reward_acc(r), r);
+                // round starts from 1 when now >= farming_start
+                let r0 = if r > 1 { r - 1 } else { 0 };
+                let farmed = self
+                    .farm_token_rates
+                    .iter()
+                    .map(|rate| U128::from(farmed_tokens(v.farmed, *rate)))
+                    .collect();
+                return Some( P4Status {
+                    stake_tokens: v.staked,
+                    stake: v.min_stake.into(),
+                    farmed_units: v.farmed.into(),
+                    farmed_tokens: farmed,
+                    boost_nfts: v.boost_nft,
+                    timestamp: self.farming_start + r0 * ROUND,
+                    total_cheddar_staked: v.total_cheddar_staked.into()
+                });
+            }
+            None => None,
+        };
+    }
+
+    // ******************* //
+    // transaction methods //
+    // ******************* //
+
+    /// withdraw NFT to a destination account using the `nft_transfer` method.
+    /// This function is considered safe and will work when contract is paused to allow user
+    /// to withdraw his NFTs.
+    #[payable]
+    pub fn withdraw_boost_nft(&mut self) {
+        assert_one_yocto();
+        let user = env::predecessor_account_id();
+        let mut vault = self.get_vault(&user);
+        self._withdraw_boost_nft(&user, &mut vault);
+    }
+
+    /// Deposit native near during the setup phase for farming rewards.
+    /// Panics when the deposit was already done or the setup is completed.
+    #[payable]
+    pub fn setup_deposit_near(&mut self) {
+        self._setup_deposit(&near(), env::attached_deposit())
+    }
+    /// FT Receiver `setup deposit` scenario
+    /// Panics on failed `unwrap()` if FT not set in `Contract.farm_tokens`
+    pub(crate) fn _setup_deposit(&mut self, token: &AccountId, amount: u128) {
+        assert!(
+            !self.setup_finalized,
+            "setup deposits must be done when contract setup is not finalized"
+        );
```

Comment:
let's assert that `token != near()`

Reply:
Why?

Reply:
because in this flow we don't support near tokens.

Reply:
Near cannot coming through `ft_transfer_call`